waspc/src/Wasp/Analyzer/Parser/Parser.hs

```diff
@@ -0,0 +1,309 @@
+-- trivia node.
+coerceStmt :: SyntaxNode -> ParserM (Maybe (WithCtx Stmt))
+coerceStmt (SyntaxNode S.Decl _ children) = do
+  startPos <- gets pstatePos
+  ((declType, declName, expr), remaining) <-
+    sequence3
+      ( coerceLexeme S.DeclType "declaration type",
+        coerceLexeme S.DeclName "declaration name",
+        coerceExpr
+      )
+      children
+  endPos <- gets pstatePos
+  mapM_ (advance . S.snodeWidth) remaining
+  return $ Just $ WithCtx (ctxFromRgn startPos endPos) (AST.Decl declType declName expr)
+coerceStmt (SyntaxNode k w _)
+  | S.syntaxKindIsTrivia k = advance w >> return Nothing
+  | otherwise = failParse $ UnexpectedNode k "in Program"
+
+-- | Try to turn CST into Expr AST. Returns @(expr, remainingNodesFromInput)@
+-- when successful.
+coerceExpr :: Action (WithCtx Expr)
+coerceExpr [] = failParse $ MissingSyntax "expression"
+coerceExpr (SyntaxNode k w children : ns)
+  | S.syntaxKindIsTrivia k = advance w >> coerceExpr ns
+  | otherwise = do
+    startPos <- gets pstatePos
+    expr <- case k of
+      S.String -> AST.StringLiteral . tail . init <$> consume w
+      S.Int -> AST.IntegerLiteral . read <$> consume w
+      S.Double -> AST.DoubleLiteral . read <$> consume w
+      S.BoolTrue -> advance w >> return (AST.BoolLiteral True)
+      S.BoolFalse -> advance w >> return (AST.BoolLiteral False)
+      S.Var -> AST.Var <$> consume w
+      S.Dict -> coerceDict children
+      S.List -> coerceList children
+      S.Tuple -> coerceTuple children
+      S.ExtImport -> coerceExtImport children
+      S.Quoter -> coerceQuoter children
+      _ -> failParse $ UnexpectedNode k "where an expression was expected"
+    endPos <- gets pstatePos
+    return (WithCtx (ctxFromRgn startPos endPos) expr, ns)
+
+coerceDict :: [SyntaxNode] -> ParserM Expr
+coerceDict syntax = do
+  (_, entries, _) <-
+    runAction syntax $
+      sequence3
+        ( coerceLexeme (S.Token T.LCurly) "{",
+          coerceEntries,
+          coerceLexeme (S.Token T.RCurly) "}"
+        )
+  return $ AST.Dict entries
+
+coerceEntries :: Action [(Identifier, WithCtx Expr)]
+coerceEntries [] = return ([], [])
+coerceEntries (n@(SyntaxNode k w children) : ns)
+  | k == S.DictEntry = (first . (:)) <$> coerceEntry children <*> coerceEntries ns
+  | k == S.Token T.Comma || S.syntaxKindIsTrivia k = advance w >> coerceEntries ns
+  | otherwise = return ([], n : ns)
+
+coerceEntry :: [SyntaxNode] -> ParserM (Identifier, WithCtx Expr)
+coerceEntry syntax = do
+  (dictKey, _, expr) <-
+    runAction syntax $
+      sequence3
+        ( coerceLexeme S.DictKey "dictionary key",
+          coerceLexeme (S.Token T.Colon) ":",
+          coerceExpr
+        )
+  return (dictKey, expr)
+
+coerceList :: [SyntaxNode] -> ParserM Expr
+coerceList syntax = do
+  (_, values, _) <-
+    runAction syntax $
+      sequence3
+        ( coerceLexeme (S.Token T.LSquare) "[",
+          coerceValues,
+          coerceLexeme (S.Token T.RSquare) "]"
+        )
```

Comment:
They are ignored for the purposes of making the `Expr`, since in a proper CST it will only be trivia nodes. But the width of each of the remaining node is used to increase the position in the state (this is what `runAction` does: uses all the remaining node to adjust the position).

Reply:
Ah ok! Could we document that a bit, even if slightly redundant? Is this documented on `runAction`?